cli/command/telemetry_utils.go

```diff
@@ -0,0 +1,172 @@
+package command
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+)
+
+type contextKey string
+
+const (
+	OtelMeterCtxKey  = contextKey("otel.meter")
+	OtelTracerCtxKey = contextKey("otel.tracer")
+)
+
+// BaseMetricAttributes returns an attribute.Set containing attributes to attach to metrics/traces
+func BaseMetricAttributes(cmd *cobra.Command) attribute.Set {
+	attrList := []attribute.KeyValue{
+		attribute.String("command.name", getCommandName(cmd)),
+	}
+	return attribute.NewSet(attrList...)
+}
+
+// InstrumentCobraCommands wraps all cobra commands' RunE funcs to set a command duration metric using otel.
+//
+// Note: this should be the last func to wrap/modify the PersistentRunE/RunE funcs before command execution.
+//
+// can also be used for spans!
+func InstrumentCobraCommands(ctx context.Context, cmd *cobra.Command, meter metric.Meter) {
+	// If PersistentPreRunE is nil, make it execute PersistentPreRun and return nil by default
+	ogPersistentPreRunE := cmd.PersistentPreRunE
+	if ogPersistentPreRunE == nil {
+		ogPersistentPreRun := cmd.PersistentPreRun
+		ogPersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+			ogPersistentPreRun(cmd, args)
+			return nil
+		}
+		cmd.PersistentPreRun = nil
+	}
+
+	// wrap RunE in PersistentPreRunE so that this operation gets executed on all children commands
+	cmd.PersistentPreRunE = func(cmd *cobra.Command, args []string) error {
+		// If RunE is nil, make it execute Run and return nil by default
+		ogRunE := cmd.RunE
+		if ogRunE == nil {
+			ogRun := cmd.Run
+			ogRunE = func(cmd *cobra.Command, args []string) error {
+				ogRun(cmd, args)
+				return nil
+			}
+			cmd.Run = nil
+		}
+		cmd.RunE = func(cmd *cobra.Command, args []string) error {
+			// insert the meter and tracer objects into the context
+			// so commands can use them in their RunE func
+			newCtx := context.WithValue(cmd.Context(), OtelMeterCtxKey, meter)
+			// newCtx := context.WithValue(newCtx, OtelTracerCtxKey, tracer))
+			cmd.SetContext(newCtx)
+
+			// start the timer as the first step of every cobra command
+			stopCobraCmdTimer := startCobraCommandTimer(cmd)
+			cmdErr := ogRunE(cmd, args)
+			stopCobraCmdTimer(cmdErr)
+			return cmdErr
+		}
+
+		return ogPersistentPreRunE(cmd, args)
+	}
+}
+
+func startCobraCommandTimer(cmd *cobra.Command) func(err error) {
+	ctx := cmd.Context()
+	meter, ok := ctx.Value(OtelMeterCtxKey).(metric.Meter)
+	if meter == nil || !ok {
+		otel.Handle(errors.New("meter not found in ctx."))
+		// return a no-op if the meter cannot be found in the command's ctx
+		return func(err error) {}
+	}
```

Comment:
Given my comment above, my suggestion is to do:

```
meter := otel.GetMeterProvider().Meter("github.com/docker/cli")
durationCounter, _ := meter.Float64Counter(...)
```

Reply:
see comment above for my doubts regarding this